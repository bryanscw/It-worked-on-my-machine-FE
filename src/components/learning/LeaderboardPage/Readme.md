```js
import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap/dist/js/bootstrap.js';

import { METHODS } from '../../../redux/ducks/apiHelper.js';
import { withReduxRouter } from '../../../utils/mock.js';

const matchObject = {
    params: {
        topicId: 1,
        levelId: 1,
    }
}
const initialState = {
    progressReducer: {
        additionalReducer: {
            leaderboard: [
                {
                "name": "Bob",
                "timing": 120000,
                },
                {
                "name": "Joe",
                "timing": 150000,
                },
                {
                "name": "Kelly",
                "timing": 200000,
                },
            ],
            isLoading: {
                [METHODS.LIST]: false
            }
        }
    }
}

withReduxRouter(<LeaderboardPage match={matchObject}/>, initialState)
```