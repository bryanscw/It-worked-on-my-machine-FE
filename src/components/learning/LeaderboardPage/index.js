import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types'

import { fetchLeaderboard, selectLeaderboardLoading, selectLeaderboardFailed, selectLeaderboard } from '../../../redux/ducks/progress';
import Loader from '../../common/Loader';

/**
 * This component displays the leaderboard for a level for a student. It is ranked based on completion time for a level.
 */
export class LeaderboardPage extends Component {
    componentDidMount() {
        const levelId = parseInt(this.props.match.params.levelId);
        this.props.fetchLeaderboard(levelId);
    }

    render() {
        const {
            match,
            leaderboardLoading,
            leaderboardFailed,
            leaderboard
        } = this.props;

        if (leaderboardLoading)
            return <Loader />;

        if (leaderboardFailed || !leaderboard)
            return <Redirect to="/not-found" />;

        console.log(leaderboard);

        const topicId = parseInt(match.params.topicId);
        const levelId = parseInt(match.params.levelId);

        // let leaderboardData = [
        //     {
        //         "name": "Bob",
        //         "timing": 10.2
        //     },
        //     {
        //         "name": "Joe",
        //         "timing": 11.2
        //     },
        //     {
        //         "name": "Caleb",
        //         "timing": 12.1
        //     },
        //     {
        //         "name": "Cathy",
        //         "timing": 13.1231
        //     },
        //     {
        //         "name": "Sarah",
        //         "timing": 14.321
        //     },
        //     {
        //         "name": "May",
        //         "timing": 15.123
        //     },
        //     {
        //         "name": "Harry",
        //         "timing": 16.143
        //     },
        //     {
        //         "name": "Matthew",
        //         "timing": 17.341
        //     },
        //     {
        //         "name": "Clyde",
        //         "timing": 19.241
        //     },
        //     {
        //         "name": "Elizabeth",
        //         "timing": 21.2321
        //     }
        // ]

        return (
            <div className="container">
                <Link className="btn btn-light mb-2" to={`/topics/${topicId}/levels/${levelId}/`}>
                    <FontAwesomeIcon icon={faChevronLeft}/> Back to Level Page
                </Link>
                <h1>Leaderboard</h1>
                <table className="table">
                    <thead className="thead-light">
                        <tr>
                            <th scope="col">Name</th>
                            <th scope="col">Timing (in minutes)</th>
                        </tr>
                    </thead>
                    <tbody>
                        {
                            leaderboard.length !== 0 && leaderboard 
                                ? (
                                    leaderboard.map(row =>
                                        <tr key={row.name}>
                                            <th scope="row">{row.name}</th>
                                            <td>{row.timing.toFixed(2)}</td>
                                        </tr>
                                    )
                                )
                                : (
                                    <tr>
                                        <td colSpan="2">No results found.</td>
                                    </tr>
                                )
                        }
                    </tbody>
                </table>
            </div>
        )
    }
}

LeaderboardPage.propTypes = {
    /** An object containing the topic ID and level ID based on which data is displayed */
   match: PropTypes.object.isRequired,

   leaderboardLoading: PropTypes.bool.isRequired,
   leaderboardFailed: PropTypes.bool,
   leaderboard: PropTypes.array.isRequired,

   fetchLeaderboard: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
    leaderboardLoading: selectLeaderboardLoading(state),
    leaderboardFailed: selectLeaderboardFailed(state),
    leaderboard: selectLeaderboard(state),
});

const dispatchers = {
    fetchLeaderboard,
};

export default connect(mapStateToProps, dispatchers)(LeaderboardPage);