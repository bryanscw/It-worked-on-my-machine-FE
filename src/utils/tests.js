import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    return <Provider store={store}>{component}</Provider>;
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    return renderWithRedux(renderWithRouter(component, route), initialState);
}