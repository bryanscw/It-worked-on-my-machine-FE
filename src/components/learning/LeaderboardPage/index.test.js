import React from 'react'
import { cleanup, waitForElement } from '@testing-library/react'
import axiosMock from 'axios'

import LeaderboardPage from './index.js';
import { renderWithReduxRouter } from '../../../utils/tests.js';

jest.mock('axios');

beforeEach(() => {
    axiosMock.get.mockResolvedValue({
        data: [
            {
                name: "Bob",
                timing: 1.2,
            },
            {
                name: "Joe",
                timing: 1.5,
            },
            {
                name: "Kelly",
                timing: 2.0,
            },
        ]
    });
});

afterEach(cleanup);

const matchObject = {
    params: {
        topicId: 1,
        levelId: 1,
    }
}

it('should take a snapshot', async () => {
    const { asFragment, getByText } = renderWithReduxRouter(<LeaderboardPage match={matchObject} />); 
    await waitForElement(() => getByText('Leaderboard'));
    expect(asFragment()).toMatchSnapshot();
});