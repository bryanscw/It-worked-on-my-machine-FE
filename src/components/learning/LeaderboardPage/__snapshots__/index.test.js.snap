// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <a
      class="btn btn-light mb-2"
      href="/topics/1/levels/1/"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-left fa-w-10 "
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
        />
      </svg>
       Back to Level Page
    </a>
    <h1>
      Leaderboard
    </h1>
    <table
      class="table"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            scope="col"
          >
            Timing
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            Bob
          </th>
          <td>
            0:00
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            Joe
          </th>
          <td>
            0:00
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            Kelly
          </th>
          <td>
            0:00
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
